File: 2019/day4/day4.go
package main

import (
	"fmt"
	"strconv"
)

func acceptable(code string) bool {
	hasDouble := false
	for i := 0; i < len(code); i++ {
		if i == len(code)-1 {
			continue
		}
		current, _ := strconv.Atoi(string(code[i]))
		next, _ := strconv.Atoi(string(code[i+1]))
		if current > next {
			return false
		} else if current == next {
			hasDouble = true
		}
	}
	if hasDouble {
		return doubleChecker(code)
	}
	return false
}

func doubleChecker(code string) bool {
	cnt := make(map[string]int)
	for _, x := range code {
		cnt[string(x)]++
	}
	hasDouble := false
	for _, v := range cnt {
		if v == 2 {
			hasDouble = true
			break
		}
	}
	return hasDouble
}

func main() {
	total := 0
	start := 231832
	end := 767346
	for i := start; i < end; i++ {
		si := strconv.Itoa(i)
		if acceptable(si) {
			total++
		}
	}
	fmt.Println(total)
}

File: 2019/day4/day4.py
#!/usr/bin/env python3

from collections import Counter

def codes(start, end):
    for x in range(start, end):
        yield str(x)

def acceptable(code):
    has_double = False
    for i, v in enumerate(code):
        if i == len(code) - 1:
            continue
        elif int(v) > int(code[i+1]):
            return False
        elif int(v) == int(code[i+1]):
            has_double = True
    if has_double:
        return double_checker(code)
    return False

def double_checker(code):
    cnt = Counter()
    for x in code:
        cnt[x] += 1
    has_double = False
    has_greater = False
    for c in cnt.most_common():
        key, value = c
        if value == 2:
            has_double = True
            break
    return has_double

#555799 pass
#699999 fail

if __name__ == '__main__': 
    total = 0
    # Range 231832-767346
    for code in codes(231832, 767346):
        if acceptable(code):
            total += 1
    print(total)